text/0031-integ-testing.md

```diff
@@ -0,0 +1,542 @@
+project, `xxx.apptest.java` for Java project, etc.
+
+This setup will enable defining a separate script that runs the apptest suite - npm script,
+Ant target, Maven phase, etc. - by using `*.apptest.*` as the file filter in your test runner.
+
+The test will be executed when the `run()` method is called on the `apptest.Test` class.
+This method will return a list of assertion failures (if any), that can then be used to notify
+the test runner of a test case failure.
+
+Test reports can be easily generated from the reporting mechanisms already available via these
+test runners.
+
+> RFC Note: See [Appendix B](#appendix-b---test-execution) for details on this design.
+
+### Testing AWS CDK Apps and construct libraries
+
+The industry standard practice is that app developers typically run pre-production instances
+of their applications or services as part of their deployment pipeline where integration tests
+are executed. This is no different for AWS CDK applications.
+
+In such cases, the apptest suite should define *ONLY* the tests and point them at the existing
+pre-production app instance, such as via a well-known endpoint or CloudFormation stack outputs.
+The apptest suite, in this case, will deploy (and towards the end destroy) a "testing" stack
+where the tests would be executed.
+
+The usage of `apptest` when testing construct libraries is different from that when it is used
+for apps.
+
+The `apptest` suite for construct libraries can be used to verify the functionality of the construct
+library when it is used with various valid configurations.
+Each `apptest` case would define a new AWS CDK `Stack` or `App` that includes the construct
+library, followed by assertions. Upon execution, both the app and the assertions will be deployed
+(and finally destroyed) as part of the test execution.
+
+### Smoke tests
+
+Smoke testing is a set of tests that are run after production deployment to reveal failures
+severe enough to rollback the deployment. Besides running tests against pre-production instances,
+apptest can be used to also write smoke test during deployments.
+
+This can be achieved by simply using apptest's assertions as part of the deployment stack.
+
+To invoke an Amazon API Gateway endpoint as part of your smoke test, simply configure -
+
+```ts
+class MyStack extends Stack {
+  constructor(scope: Construct, id: string) {
+    // ...
+    const restApi = new apigateway.RestApi(this, 'MyRestApi');
+    // ...
+    new HttpInvoke(this, 'SmokeTest', {
+      endpoint: `${restApi.url}/ping`,
+    })
+  }
+}
+```
+
+When the smoke test fails, the CloudFormation stack deployment will fail which triggers a rollback.
+
+## Working Backwards - README.md
+
+### `cdk-deploy`
+
+This jsii module provides the ability to programmatically deploy and destroy AWS CDK apps.
+
+A CDK app can be deployed by calling -
+
+```ts
+import { App } from 'cdk-deploy';
+
+const app = new App('/path/to/cdk/app');
```

Comment:
Is this a path to a cloud assembly or to a directory with `cdk.json`? Kind of feels like it should accept `cdk.App` instance instead, no?